### ISSUE TITLE
Argument names are significant in overloads?

### ISSUE THREAD

**Issue by erikkemperman** [python/mypy#8073]
I'm not sure if this is a bug or a deliberate design choice -- but if it's the latter I did not encounter any description of it in the documentation, and the error messages provided no clues (or not to me, at any rate).

I was expecting the following minimal repro to pass `mypy --strict` (I'm on version 0.750)

```python3
@overload
def fun(i: int) -> str:
    ...

@overload
def fun(s: str) -> int:
    ...

def fun(x: Union[int, str]) -> Union[int, str]:
    if isinstance(x, str):
        return int(x)
    return str(x)
```

However this complains that *Overloaded function implementation does not accept all possible arguments of signature 1* (and likewise for signature 2).

Now, if I change the argument names in the stubs to be `x` just like in the implementation, suddenly `mypy --strict` is happy with it.

```python3
@overload
def fun(x: int) -> str:
    ...

@overload
def fun(x: str) -> int:
    ...

def fun(x: Union[int, str]) -> Union[str, int]:
    if isinstance(x, str):
        return int(x)
    return str(x)
```

So it appears that the checker takes the argument names into consideration when it reasons about my signatures? For some reason I expected it to only look at position and type.

If that is the case, perhaps the error message and/or the documentation ought to mention this? I can imagine it might be common for people to want to name their arguments differently for various types, which apparently is not what mypy expects for overloads.

Perhaps it should simply warn about mismatched argument names in a separate error message?

Apologies if this is a duplicate, I searched for a bit but nothing obvious came up.

Also, thanks a lot for this great addition to the Python ecosystem!

**Comment by JelleZijlstra** [comment]
The error message could definitely be improved, but just to be clear mypy is right here: a call like `fun(i=3)` would be accepted by the first overload, but not by the implementation.

**Comment by erikkemperman** [comment]
@JelleZijlstra Thank you for the quick response! Ah, yes, I wasn't thinking about these as keyword arguments, but of course you're absolute right. And in fact, if I use the shiny new positional-only marker then it actually works even with different argument names:

```python3
@overload
def fun(i: int, /) -> str:
    ...

@overload
def fun(s: str, /) -> int:
    ...

def fun(x: Union[int, str], /) -> Union[int, str]:
    if isinstance(x, int):
        return str(x)
    return int(x)
```

So definitely not a bug, although perhaps the error message could be improved a bit.

I guess I should close this issue, if anyone else makes the same mistake hopefully they will find this and be spared some confusion.

Thanks again!

**Comment by tueda** [comment]
Is there a right way to overload functions with different names as in this example *without* using positional-only arguments (which requires Python 3.8)?

I tried to modify the implementation of `fun` in such a way that keyword arguments are accepted:
```python
from typing import Optional, Union, overload


@overload
def fun(i: int) -> str:
    ...


@overload
def fun(s: str) -> int:
    ...


def fun(
    x: Optional[Union[int, str]] = None, *, i: Optional[int] = None, s: Optional[str] = None
) -> Union[int, str]:
    if i is not None:
        assert x is None
        assert isinstance(i, int)
        assert s is None
        return fun(i)

    if s is not None:
        assert x is None
        assert i is None
        assert isinstance(s, str)
        return fun(s)

    if isinstance(x, str):
        return int(x)
    return str(x)

# The following calls correctly work.
print(repr(fun(10)))
print(repr(fun('10')))
print(repr(fun(i=10)))
print(repr(fun(s='10')))
```
but still I get
```
main.py:14: error: Overloaded function implementation does not accept all possible arguments of signature 1
main.py:14: error: Overloaded function implementation does not accept all possible arguments of signature 2
```

**Comment by JelleZijlstra** [comment]
You can mark the arguments as positional-only to the typechecker by prefixing their names with `__`; see python/peps#147.

**Comment by tueda** [comment]
Thank you for the prompt answer! I have confirmed that mypy doesn't complain about the following code:

```python
from typing import Optional, Union, overload


@overload
def fun(__i: int) -> str:
    ...


@overload
def fun(__s: str) -> int:
    ...


def fun(x: Optional[Union[int, str]]) -> Union[int, str]:
    if isinstance(x, str):
        return int(x)
    return str(x)


# The following calls correctly work.
print(repr(fun(10)))
print(repr(fun("10")))
```
